### PR TITLE
feat: support SASL SCRAM mechanism

### PR DESCRIPTION
**Pull request by WenyXu**
#239

support SASL SCRAM-SHA-256 and SCRAM-SHA-512

- [x] I've read the contributing section of the project [CONTRIBUTING.md](https://github.com/influxdata/rskafka/blob/main/CONTRIBUTING.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

**Comment by WenyXu** [comment]
> I would like to wait for upstream to merge and release this feature. I agree with the general approach of using a dedicated crate instead of implementing all these algorithms within rskafka. 👍

Now, we can push forward.

File: src/messenger.rs [hunk 1]
```diff
@@ -549,6 +557,63 @@ where
...
+
+        let Credentials { username, password } = config.credentials();
+        let config = SASLConfig::with_credentials(None, username, password).unwrap();
+        let sasl = rsasl::prelude::SASLClient::new(config);
+        let raw_mechanisms = resp.mechanisms.0.unwrap_or_default();
+        let mechanisms = raw_mechanisms
+            .iter()
+            .map(|mech| Mechname::parse(mech.0.as_bytes()).map_err(SaslError::InvalidSaslMechanism))
+            .collect::<Result<Vec<_>, SaslError>>()?;
+        debug!("Supported mechanisms {:?}", mechanisms);
```
Comment (crepererum) [inline src/messenger.rs:585]:
```suggestion
        debug!(?mechanisms, "Supported SASL mechanisms {:?}");
```

File: src/messenger.rs [hunk 2]
```diff
@@ -549,6 +557,63 @@ where
...
+        let prefer_mechanism =
+            Mechname::parse(mechanism.as_bytes()).map_err(SaslError::InvalidSaslMechanism)?;
+        if !mechanisms.contains(&prefer_mechanism) {
+            return Err(SaslError::UnsupportedSaslMechanism);
+        }
+        let mut session = sasl
+            .start_suggested(&[prefer_mechanism])
+            .map_err(|_| SaslError::UnsupportedSaslMechanism)?;
+        // let selected_mechanism = session.get_mechname();
+        debug!("Using {:?} for the SASL Mechanism", mechanism);
```
Comment (crepererum) [inline src/messenger.rs:595]:
```suggestion
        debug!(?mechanism, "Using SASL Mechanism");
```

File: src/messenger.rs [hunk 3]
```diff
@@ -549,6 +557,63 @@ where
...
+                .expect("step errored!");
+
+            data = Some(out.into_inner());
+
+            // Returns `true` if step needs to be called again with another batch of data.
+            state.is_running()
+        } {
+            let authentication_response = self.sasl_authentication(data.take().unwrap()).await?;
+            data = Some(authentication_response.auth_bytes.0);
+        }
```
Comment (crepererum) [inline src/messenger.rs:615]:
```suggestion
        // we step through the auth process, starting on our side with NO data received so far
        let mut data_received: Option<Vec<u8>> = None;

        loop {
            let mut to_sent = Cursor::new(Vec::new());
            let state = session
                .step(data_received.as_deref(), &mut to_sent)?;

            if !state.is_running() {
                break;
            }
            
            let authentication_response = self.sasl_authentication(to_sent.into_inner().await?;
            data_received = Some(authentication_response.auth_bytes.0);
        }
```

Some improvements:

- use slightly more readable loop construct
- separate variables for read and write direction (which also eliminates an unwrap)
- replace `except` with proper error handling (you probably need to extend `SaslError`)
- remove the code comment that mentions that you've copied this code from some example :wink:

**Review by crepererum — CHANGES_REQUESTED** [review]
Some nitpicks:

- structured logging (which is esp. helpful if you never yeet through production logs trying to debug why your auth suddenly broke)
- some general recommendations for the the auth stepping

File: src/messenger.rs (hunk 3 above)
Comment (WenyXu) [reply src/messenger.rs]: Resolved in [01f1107](https://github.com/influxdata/rskafka/pull/247/commits/01f1107834d6fb3f116e027ae67625e7eb1a9282)

**Review by crepererum — APPROVED** [review]

**Pull request merged**